liquidator/liquidator.js

```diff
@@ -158,12 +168,54 @@ class Liquidator {
       // Note: query the time again during each iteration to ensure the deadline is set reasonably.
       const currentBlockTime = this.empClient.getLastUpdateTime();
 
+      // Calculate the amount of tokens we will attempt to liquidate.
+      let tokensToLiquidate;
+
+      // If the user specifies `maxTokensToLiquidateWei`, then we must make sure that it follows an important constraint:
+      // we cannot bring the position down below the `minSponsorPosition`.
+      if (maxTokensToLiquidateWei) {
+        // First, we check if `maxTokensToLiquidateWei > tokensOutstanding`, if so then we'll liquidate the full position.
+        if (this.toBN(maxTokensToLiquidateWei).gte(this.toBN(position.numTokens))) {
+          tokensToLiquidate = this.toBN(position.numTokens);
+        } else {
+          // If we're not liquidating the full position, then we cannot liquidate the position below the `minSponsorTokens` constraint.
+          // `positionTokensAboveMinimum` is the maximum amount of tokens any liquidator can liquidate while taking `minSponsorTokens` into account.
+          const positionTokensAboveMinimum = this.toBN(position.numTokens).sub(this.toBN(this.empMinSponsorSize));
+
+          // Finally, we cannot liquidate more than `maxTokensToLiquidate`.
+          tokensToLiquidate = this.BN.min(positionTokensAboveMinimum, this.toBN(maxTokensToLiquidateWei));
+        }
```

Comment:
I know we've discussed this a bunch, but I still think the bot will be silent if it sends a partial liquidation. That seems like an issue. To be more concrete, let's consider the following situation:
1. minSponsorTokens is 0 (just to make the math simpler). This means that you liquidate min(maxTokensToLiquidateWei, position.numTokens).
2. Let's say that position.numTokens is larger than maxTokensToLiquidateWei. This means that we're sending out a partial liquidation because we don't have the funds to pay for the full liquidation.
3. By my estimation, we will not send out an error log/alert when that happens.

Before this change, we would send out an error if we didn't have the funds to fully liquidate a position. Now, we send a partial liquidation (which is good), but we no longer send an alert if there aren't enough funds to fully liquidate the position, which seems like something we don't want (at least for the liquidation bots we run).

Does that make sense? Am I reading the code right?

Reply:
You're correct. So do you think this should be an error message? an info? 

I agree we might want to know if a partial liquidation happens. 

If we were to add in this error message, we would have 3 different errors for liquidations:

1. The position size is at the minimum sponsor size and the liquidator bot doesn't have enough `maxTokensToLiquidateWei` to liquidate the entire position.
2. The liquidator bot is attempting a partial liquidation.
3. The liquidator bot attempted a liquidation but it failed due to insufficient allowance/balance

I can change the error messages to reflect this. We also might want to consolidate these cases so I'm curious what you think.

Reply:
I could see it being configurable or an error. I think, internally, we definitely want this to be an error message so it alerts us loudly and immediately.

I think consolidating them would be great if possible. I think at least 1 and 2 are pretty much the same thing (as far as the error is concerned) -- not enough balance to do a full liquidation. For 3, the cause _could_ be different, so we might want it to be separate. Not super opinionated here, though.

Reply:
The difference between 1 and 2 is that we cannot liquidate if we get to situation 1, but we will submit a partial liquidation for 2.

Reply:
Sorry, what I meant to say was that they have the same cause (low balance), so the error would basically read “not enough balance to fully liquidate” or something like that in both cases.

Reply:
Agreed. I updated the logs and unit tests (comments below) to reflect this suggestion